crates/sequencer/src/mempool_syncer.rs

```diff
@@ -0,0 +1,156 @@
+use tracing::{debug, error, info, instrument};
+
+#[derive(Clone)]
+pub struct MempoolSyncer<DB>
+where
+    DB: SequencerLedgerOps + Clone + Send + Sync + 'static,
+{
+    /// Database for ledger operations
+    ledger_db: DB,
+    /// Buffer for mempool transactions before storing into the ledger db
+    /// Mapping: Tx hash to rlp encoded transaction
+    transactions_buffer: Arc<Mutex<HashMap<Vec<u8>, Vec<u8>>>>,
+    /// Transaction hashes to be removed from mempool ledger db
+    transactions_to_remove_buffer: Arc<Mutex<HashSet<Vec<u8>>>>,
+    /// sequencer websocket endpoint
+    sequencer_ws_endpoint: String,
+}
+
+impl<DB> MempoolSyncer<DB>
+where
+    DB: SequencerLedgerOps + Clone + Send + Sync + 'static,
+{
+    /// Creates a new MempoolSyncer
+    pub fn new(ledger_db: DB, sequencer_ws_endpoint: String) -> Self {
+        Self {
+            ledger_db,
+            transactions_buffer: Arc::new(Mutex::new(HashMap::new())),
+            transactions_to_remove_buffer: Arc::new(Mutex::new(HashSet::new())),
+            sequencer_ws_endpoint,
+        }
+    }
+
+    #[instrument(name = "MempoolSyncer", skip_all)]
+    pub async fn run(self, shutdown_signal: GracefulShutdown) {
+        let this = self.clone();
+        tokio::spawn(async move {
+            // Start the subscription task
+            this.run_subscription_task().await;
+        });
+
+        tokio::spawn(async move {
+            // Start the update task
+            self.update_mempool_transaction_db_task().await;
+        });
+
+        let _ = shutdown_signal.await;
+        info!("Shutting down mempool syncer");
+    }
+
+    pub async fn run_subscription_task(&self) {
+        loop {
+            let exponential_backoff = ExponentialBackoff::default();
+            let _ = retry_backoff(exponential_backoff, || async {
+                subscribe_to_mempool_transaction_updates(
+                    &self.sequencer_ws_endpoint,
+                    self.transactions_buffer.clone(),
+                    self.transactions_to_remove_buffer.clone(),
+                )
+                .await
+                .map_err(|e| {
+                    error!("Subscription error: {}", e);
+                    backoff::Error::Transient {
+                        err: e,
+                        retry_after: None,
+                    }
+                })
+            })
+            .await;
+        }
+    }
+
+    pub async fn update_mempool_transaction_db_task(&self) {
+        loop {
+            // Waiting at least 2 seconds here so that txs that got in block are removed so we do less db ops
+            tokio::time::sleep(Duration::from_secs(3)).await;
+            {
+                let transactions_buffer = self.transactions_buffer.clone();
+                let mut txs_buffer = transactions_buffer.lock();
+
+                let txs = txs_buffer.drain().collect();
+                if let Err(e) = self.ledger_db.batch_insert_mempool_txs(txs) {
```

Comment:
Just wondering why save the transactions into mempool specific tables while we can just insert them directly to the listen mode mempool. When the listen mode is converted into an actual sequencer, the transactions will be there.